Core/include/Acts/TrackFitting/MbfSmoother.hpp

```diff
@@ -0,0 +1,138 @@
+                                  false>::Covariance;
+    using Parameters =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Parameters;
+    using Covariance =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Covariance;
+
+    // This is used to build a covariance matrix view in the .cpp file
+    unsigned int calibratedSize{0};
+    const double* calibrated{nullptr};
+    const double* calibratedCovariance{nullptr};
+    Projector projector;
+
+    Jacobian jacobian{nullptr};
+
+    Parameters predicted{nullptr};
+    Covariance predictedCovariance{nullptr};
+    Parameters filtered{nullptr};
+    Covariance filteredCovariance{nullptr};
+    Parameters smoothed{nullptr};
+    Covariance smoothedCovariance{nullptr};
+
+    InternalTrackState() = default;
+
+    template <typename TrackStateProxy>
+    explicit InternalTrackState(TrackStateProxy ts)
+        : jacobian(ts.jacobian()),
+          predicted(ts.predicted()),
+          predictedCovariance(ts.predictedCovariance()),
+          filtered(ts.filtered()),
+          filteredCovariance(ts.filteredCovariance()),
+          smoothed(ts.smoothed()),
+          smoothedCovariance(ts.smoothedCovariance()) {
+      if (ts.typeFlags().test(TrackStateFlag::MeasurementFlag)) {
+        calibratedSize = ts.calibratedSize();
+        // Note that we pass raw pointers here which are used in the correct
+        // shape later
+        calibrated = {ts.effectiveCalibrated().data()};
+        calibratedCovariance = {ts.effectiveCalibratedCovariance().data()};
+        projector = {ts.projector()};
+      }
+    }
+  };
+
+  /// Run the Kalman smoothing for one trajectory.
+  ///
+  /// @param[in] gctx The geometry context to be used
+  /// @param[in,out] trajectory The trajectory to be smoothed
+  /// @param[in] entryIndex The index of state to start the smoothing
+  /// @param[in] logger Where to write logging information to
+  template <typename traj_t>
+  Result<void> operator()(const GeometryContext& gctx, traj_t& trajectory,
+                          std::size_t entryIndex,
+                          const Logger& logger = getDummyLogger()) const {
+    (void)gctx;
+    (void)logger;
+
+    using TrackStateProxy = typename traj_t::TrackStateProxy;
+
+    TrackStateProxy start_ts = trajectory.getTrackState(entryIndex);
+
+    BoundMatrix big_lambda_hat = BoundMatrix::Zero();
+    BoundVector small_lambda_hat = BoundVector::Zero();
+
+    trajectory.applyBackwards(start_ts.index(), [&](TrackStateProxy ts) {
+      // ensure the track state has a smoothed component
+      ts.addComponents(TrackStatePropMask::Smoothed);
+
+      InternalTrackState internalTrackState(ts);
+
+      calculateSmoothed(internalTrackState, big_lambda_hat, small_lambda_hat);
+
+      if (!ts.hasPrevious()) {
```

Comment:
should this check be moved at the beginning of the `applyBackwards` (to L105) to avoid unnecessary operations in case `ts` doesn't have previous state?

Reply:
It's important that `calculateSmoothed` is called before we exit. The reason being that this is the last state of the track so we do not transport the magic numbers to another state. I can clarify with a code comment